File: src/main/java/io/github/aalsanie/codes/OutcomeException.java
package io.github.aalsanie.codes;

import java.util.Objects;

public final class OutcomeException extends RuntimeException {
    private final Outcome outcome;

    public OutcomeException(Outcome outcome) {
        this(outcome, null);
    }

    public OutcomeException(Outcome outcome, Throwable cause) {
        super(requireFailed(outcome).getMessage(), cause);
        this.outcome = outcome;
    }

    public Outcome getOutcome() {
        return outcome;
    }

    private static Outcome requireFailed(Outcome outcome) {
        Objects.requireNonNull(outcome, "outcome");
        if (!outcome.isFailed()) {
            throw new IllegalArgumentException("only failed outcomes can be converted to OutcomeException");
        }
        return outcome;
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/OutcomeException.kt
@file:JvmName("OutcomeExceptions")

package io.github.aalsanie.codes

class OutcomeException
    @JvmOverloads
    constructor(
        val outcome: Outcome,
        cause: Throwable? = null,
    ) : RuntimeException(outcome.message, cause) {
        init {
            require(outcome.isFailed) {
                "only failed outcomes can be converted to OutcomeException"
            }
        }
    }

@JvmOverloads
fun Outcome.toException(cause: Throwable? = null): OutcomeException = OutcomeException(this, cause)
